File: scala/src/RandomGraphGenerator.scala
import base.Vertex

import scala.collection.immutable.HashSet
import scala.collection.mutable
import scala.util.Random

object RandomGraphGenerator {
  def random_choice(rng: Random, vseq: mutable.ArrayBuffer[Vertex]): Vertex = {
    val i = rng.nextInt(vseq.size)
    vseq(i)
  }

  def random_sample(rng: Random, vset: mutable.Set[Vertex]): Vertex = {
    val i = rng.nextInt(vset.size)
    vset.iterator.drop(i).next()
  }

  def remove_from(vseq: mutable.Buffer[Vertex], v: Vertex): Unit = {
    val i = vseq.indexOf(v)
    vseq(i) = vseq.last
    vseq.trimEnd(1)
  }

  def new_adjacencies(n: Int): mutable.IndexedSeq[mutable.Set[Vertex]] = {
    mutable.IndexedSeq.fill(n) { mutable.Set() }
  }

  def new_undirected(rng: Random, order: Int, size: Int): UndirectedGraph = {
    require(order > 0)
    val fully_meshed_size = order
      .asInstanceOf[Long] * (order.asInstanceOf[Long] - 1) / 2
    require(
      size <= fully_meshed_size,
      f"$order nodes accommodate at most $fully_meshed_size edges"
    )
    val unsaturated_vertices: mutable.ArrayBuffer[Vertex] =
      mutable.ArrayBuffer(0 until order: _*)
    val adjacency_sets = new_adjacencies(order)
    val adjacency_complements = new_adjacencies(order)
    for (_ <- 1 to size) {
      assert(
        unsaturated_vertices
          .forall(v => adjacency_sets(v).size < order - 1),
        adjacency_sets
      )
      val v = random_choice(rng, unsaturated_vertices)
      val w =
        if (adjacency_complements(v).nonEmpty) {
          random_sample(rng, adjacency_complements(v))
        } else {
          var w = v
          while (w == v || adjacency_sets(v).contains(w)) {
            w = random_choice(rng, unsaturated_vertices)
          }
          w
        }
      assert(v != w)
      assert(!adjacency_sets(v).contains(w))
      assert(!adjacency_sets(w).contains(v))
      for ((x, y) <- Array((v, w), (w, v))) {
        adjacency_sets(x) += y
        val neighbours = adjacency_sets(x).size
        if (neighbours == order - 1) {
          remove_from(unsaturated_vertices, x)
          adjacency_complements(x).clear()
        } else if (neighbours == order / 2) {
          // start using adjacency complement
          require(adjacency_complements(x).isEmpty)
          adjacency_complements(x) ++= unsaturated_vertices.filter(_ != x)
          adjacency_complements(x) --= adjacency_sets(x)
          assert(!adjacency_complements(x).contains(x))
        } else if (neighbours > order / 2) {
          val ok = adjacency_complements(x).remove(y)
          require(ok)
        }
      }
    }
    val adjacencies = IndexedSeq.empty[Set[Vertex]] ++ adjacency_sets
      .map(neighbours => new HashSet[Vertex] ++ neighbours)
    val g = new SlimUndirectedGraph(adjacencies)
    require(g.order == order)
    require(g.size == size)
    g
  }
}

File: java/src/main/java/be/steinsomers/bron_kerbosch/study/RandomGraphGenerator.java
package be.steinsomers.bron_kerbosch.study;

import be.steinsomers.bron_kerbosch.UndirectedGraph;
import be.steinsomers.bron_kerbosch.util;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

@SuppressWarnings("CollectionWithoutInitialCapacity")
final class RandomGraphGenerator {
    private static List<Set<Integer>> new_sets(int n) {
        return Stream.generate(() -> new HashSet<Integer>()).limit(n).collect(Collectors.toList());
    }

    private final Random rng;

    RandomGraphGenerator(Random rng) {
        this.rng = rng;
    }

    private int pickNewNeighbour(int v) {
        var ac = adjacencyComplements.get(v);
        if (ac.isEmpty()) {
            // not yet using adjacency complement for this vertex
            var neighbours = adjacencySets.get(v);
            int w = v;
            while (w == v || neighbours.contains(w)) {
                w = util.RandomChoice(rng, unsaturatedVertices);
            }
            return w;
        } else {
            return util.RandomSample(rng, ac);
        }
    }

    private void addEdge(int v, int w) {
        final var order = adjacencySets.size();
        assert order / 2 < order - 1; // for readability and otherwise nothing to randomize anyways
        var neighbours = adjacencySets.get(v);
        neighbours.add(w);
        if (neighbours.size() >= order / 2) {
            var ac = adjacencyComplements.get(v);
            if (neighbours.size() == order / 2) {
                // start using adjacency complement
                assert ac.isEmpty();
                unsaturatedVertices.stream()
                        .filter(u -> u != v && !neighbours.contains(u))
                        .forEach(ac::add);
            } else if (neighbours.size() < order - 1) {
                // continue using adjacency complement
                var ok = ac.remove(w);
                assert ok;
            } else {
                // stop using vertex entirely
                util.RemoveFrom(unsaturatedVertices, v);
                ac.clear(); // clean up, may help or harm performance
            }
        }
    }

    private ArrayList<Integer> unsaturatedVertices;
    private List<Set<Integer>> adjacencySets;
    private List<Set<Integer>> adjacencyComplements;

    UndirectedGraph newUndirected(int order, int size) {
        assert order > 2;
        assert size >= 0;
        var fullyMeshedSize = ((long) order) * (order - 1) / 2;
        if (size > fullyMeshedSize) {
            throw new IllegalArgumentException(String.format(
                    "%d nodes accommodate at most %d edges", order, fullyMeshedSize));
        }

        unsaturatedVertices = IntStream.range(0, order).boxed()
                .collect(Collectors.toCollection(ArrayList::new));
        adjacencySets = new_sets(order);
        adjacencyComplements = new_sets(order);
        for (int i = 0; i < size; ++i) {
            assert unsaturatedVertices.stream()
                    .allMatch(v -> adjacencySets.get(v).size() < order - 1);
            int v = util.RandomChoice(rng, unsaturatedVertices);
            int w = pickNewNeighbour(v);
            assert v != w;
            assert !adjacencySets.get(v).contains(w);
            assert !adjacencySets.get(w).contains(v);
            addEdge(v, w);
            addEdge(w, v);
        }
        var adjacencies = adjacencySets.stream().map(Set::copyOf).collect(Collectors.toList());
        var g = new UndirectedGraph(adjacencies);
        if (g.order() != order) throw new AssertionError("order mishap");
        if (g.size() != size) throw new AssertionError("size mishap");
        return g;
    }
}
